src/selfplay/tournament.cc

```diff
@@ -57,13 +59,244 @@ const OptionId kTrainingId{
+    } else {
+      m = Move(BoardSquare(0, 4), BoardSquare(0, 6));
+    }
+    return m;
+  }
+  if (p != 0) idx++;
+  // Formats e4 1e5 de5 d1e5 - with optional x's - followed by =Q for
+  // promotions, and even more characters after that also optional.
+  int r1 = -1;
+  int c1 = -1;
+  int r2 = -1;
+  int c2 = -1;
+  int p2 = -1;
+  bool pPending = false;
+  for (; idx < san.size(); idx++) {
+    if (san[idx] == 'x') continue;
+    if (san[idx] == '=') {
+      pPending = true;
+      continue;
+    }
+    if (san[idx] >= '1' && san[idx] <= '9') {
+      r1 = r2;
+      r2 = san[idx] - '1';
+      continue;
+    }
+    if (san[idx] >= 'a' && san[idx] <= 'h') {
+      c1 = c2;
+      c2 = san[idx] - 'a';
+      continue;
+    }
+    if (pPending) {
+      if (san[idx] == 'Q') {
+        p2 = 2;
+      } else if (san[idx] == 'B') {
+        p2 = 3;
+      } else if (san[idx] == 'N') {
+        p2 = 4;
+      } else if (san[idx] == 'R') {
+        p2 = 5;
+      }
+      pPending = false;
+      break;
+    }
+    break;
+  }
+  if (r1 == -1 || c1 == -1) {
+    // Need to find the from cell based on piece.
+    int sr1 = r1;
+    int sr2 = r2;
+    if (board.flipped()) {
+      if (sr1 != -1) sr1 = 7 - sr1;
+      sr2 = 7 - sr2;
+    }
+    BitBoard searchBits;
+    if (p == 0) {
+      searchBits = (board.pawns() & board.ours());
+    } else if (p == 1) {
+      searchBits = board.our_king();
+    } else if (p == 2) {
+      searchBits = (board.queens() & board.ours());
+    } else if (p == 3) {
+      searchBits = (board.bishops() & board.ours());
+    } else if (p == 4) {
+      searchBits = board.our_knights();
+    } else if (p == 5) {
+      searchBits = (board.rooks() & board.ours());
+    }
+    auto plm = board.GenerateLegalMoves();
+    int pr1 = -1;
+    int pc1 = -1;
+    for (BoardSquare sq : searchBits) {
+      if (sr1 != -1 && sq.row() != sr1) continue;
+      if (c1 != -1 && sq.col() != c1) continue;
+      if (std::find(plm.begin(), plm.end(),
+                    MoveFor(sq.row(), sq.col(), sr2, c2, p2)) == plm.end())
+        continue;
+      if (pc1 != -1) {
+        std::cerr << "Ambiguous!!" << std::endl;
+        throw Exception("Opening book move seems ambiguous.");
+      }
+      pr1 = sq.row();
+      pc1 = sq.col();
+    }
+    if (pc1 == -1) {
+      std::cerr << "No Match!!" << std::endl;
+      throw Exception("Opening book move seems illegal.");
+    }
+    r1 = pr1;
+    c1 = pc1;
+    if (board.flipped()) {
+      r1 = 7 - r1;
+    }
+  }
+  Move m = MoveFor(r1, c1, r2, c2, p2);
+  if (board.flipped()) m.Mirror();
+  return m;
+}
+
+std::vector<MoveList> ReadBook(const std::string& path) {
+  std::vector<MoveList> result;
+  MoveList cur;
+  ChessBoard board;
+  board.SetFromFen(ChessBoard::kStartposFen);
+  std::ifstream file(path);
+  std::string line;
+  bool in_comment = false;
+  while (std::getline(file, line)) {
+    if (line.empty() || line[0] == '[') {
```

Comment:
It is true that I've not written support for 'full import quality pgn' - probably only a slightly relaxed export quality support. On the other hand new lines are not entirely arbitrary - blank lines have meaning, so you can't just discard whitespace tokens.

I think I'd rather not spend time on a formal parser of informal quality inputs, can leave that to someone to do as a follow up if they are so inclined.

Reply:
The PGN spec curiously includes a grammar, that fails to mention some of the things mentioned in the body text - so that is fun...
I've not addressed this, although I did leave a TODO about handling tags with newlines - although I suspect there are other more important deficiencies (like not ignoring NAGs properly perhaps - handling move number indicators without any . in them, or with more than one . ... :P)